### ISSUE TITLE
Tests error out after update from 9.18.1 to 10.1.1

### ISSUE THREAD

**Issue by mmoll** [pact-foundation/pact-js#926]
### Software versions

_Please provide at least OS and version of pact-js_

- **OS**:  Linux
- **Consumer Pact library**: Pact JS 10.1.1
- **Provider Pact library**: not relevant
- **Node Version**:  v18.7.0

### Issue Checklist

Please confirm the following:

- [x] I have upgraded to the latest
- [x] I have the read the FAQs in the Readme
- [ ] I have triple checked, that there are **no unhandled promises** in my code and have [read](https://github.com/pact-foundation/pact-js/blob/master/docs/troubleshooting.md#test-fails-when-it-should-pass) the section on intermittent test failures
- [x] I have set my log level to debug and attached a log file showing the complete request/response cycle
- [ ] For bonus points and virtual high fives, I have created a reproduceable git repository (see below) to illustrate the problem

### Expected behaviour

Pact JS 10.1.1 should be compatible with 9.18.1 and everything should work as-is, when not using V3 Matchers/Pacts.

### Actual behaviour

After update 9.18.1 -> 10.1.1, tests stop to work.

### Steps to reproduce

The following works fine with 9.18.1:

```typescript
import { Matchers, Pact } from "@pact-foundation/pact";
import { fetchNavigation } from "../../../utils/fetcher";

const { extractPayload, like } = Matchers;

const mockSearchProvider = new Pact({
  consumer: "my_consumer",
  provider: "my_service",
  host: "localhost",
  port: 9998,
  logLevel: "debug",
});

jest.mock("next/config", () => ({
  __esModule: true,
  default: () => ({
    serverRuntimeConfig: {
      myServiceBaseUrl: "http://localhost:9998",
    },
  }),
}));

const acceptJson = "application/json";
const headers = {
  Accept: acceptJson,
};

describe("Categories API Pact test", () => {
  beforeAll(() => mockSearchProvider.setup());
  afterEach(() => mockSearchProvider.verify());
  afterAll(() => mockSearchProvider.finalize());

  describe("get categories", () => {
    test("fetch all categories for navigation", async () => {
      const categoriesExpectation = like([]);
      await mockSearchProvider.addInteraction({
        state: "all categories",
        uponReceiving: "a request to get categories",
        withRequest: {
          method: "GET",
          path: "/categories",
          headers: headers,
        },
        willRespondWith: {
          status: 200,
          body: categoriesExpectation,
        },
      });

      const category = await fetchNavigation("de_DE");

      expect(category).toStrictEqual(extractPayload(categoriesExpectation));
    });
  });
});
```

result with 10.1.1:

```[2022-08-16 18:30:49.498 +0000] DEBUG (500440 on msail): pact-core@13.7.7: Initalising native core at log level 'debug'
[2022-08-16 18:30:49.514 +0000] DEBUG (500440 on msail): pact@10.1.1: Setting up Pact mock server with Consumer "my_consumer" and Provider "my_service"
        using mock service on Port: "9998"
[2022-08-16 18:30:49.514 +0000] DEBUG (500440 on msail): pact@10.1.1: mock service started on port: 65531
 FAIL  tests/jest/pact/categories.spec.ts
  Categories API Pact test
    get categories
      ✕ fetch all categories for navigation (28 ms)

  ● Categories API Pact test › get categories › fetch all categories for navigation

    FetchError: request to http://localhost:9998/categories failed, reason: connect ECONNREFUSED 127.0.0.1:9998

      at ClientRequest.<anonymous> (node_modules/node-fetch/lib/index.js:1491:11)

  ● Categories API Pact test › get categories › fetch all categories for navigation

    Error in native callback

      31 | describe("Categories API Pact test", () => {
      32 |   beforeAll(() => mockSearchProvider.setup());
    > 33 |   afterEach(() => mockSearchProvider.verify());
         | ^
      34 |   afterAll(() => mockSearchProvider.finalize());
      35 |
      36 |   describe("get categories", () => {

      at Object.mockServerMismatches (node_modules/@pact-foundation/pact-core/src/consumer/index.ts:94:13)
      at Pact.Object.<anonymous>.Pact.verify (node_modules/@pact-foundation/src/httpPact/index.ts:202:39)
      at Object.<anonymous> (tests/jest/pact/categories.spec.ts:33:1)
```

Note port 9998 vs. 65531. However, even after changing the port to 65531 just to get things going, the same error can be seen.

**Comment by mefellows** [comment]
I would expect to see a lot more logging at debug level - could you please share that?

> FetchError: request to http://localhost:9998/categories failed, reason: connect ECONNREFUSED 127.0.0.1:9998

It's possible that the mock server is binding to an IP that is not `127.0.0.1` but your HTTP client thinks localhost is that IP. You could update the host in the config to explicitly set the IP to `127.0.0.1` (or another as appropriate) and see if that helps.

**Comment by mefellows** [comment]
OK just looking at the code it seems to be ignoring your request for port `9998`, so I don't think the above guess is on track. One possibility is that the port is unavailable, and somewhere along the line we detect that and start on a new one. I can't see obviously where, but I suspect getting the debug level logs will help.

**Comment by mmoll** [comment]
> I would expect to see a lot more logging at debug level - could you please share that?

This is all I'm getting. Also with 10.x, `logs/pact.log` is not getting anything appended... is there some other place that needs the log level increased?

**Comment by mefellows** [comment]
Could you please share a repro project I can clone and test? We don't need your work code, just a minimal example that illuminates the problem.

**Comment by mmoll** [comment]
> Could you please share a repro project I can clone and test?

Sure thing, https://github.com/mmoll/pact_js_update - a1114ae303a7378f2f96e649a985e124c5b7781d works fine and dc30219da6adda0740462b2769dd0dc35630498e is showing the problem

**Comment by YOU54F** [comment]
Hey @mmoll 

Just tested with your example

In the original commit, I get an error on my mac, due to a port conflict, so cannot generate a pact file

https://github.com/mmoll/pact_js_update/commit/a1114ae303a7378f2f96e649a985e124c5b7781d#diff-d9b29b573a844a032fa2ce5b1bf9658b0617053ba23a10a82ad5eca13d718912R10-R18

This error is the same on both commits.

```
    Cannot open port 9998 on ipv4 or ipv6 interfaces

      at node_modules/@pact-foundation/src/common/net.ts:44:13
      at tryCatcher (node_modules/bluebird/js/release/util.js:16:23)
      at Promise._settlePromiseFromHandler (node_modules/bluebird/js/release/promise.js:517:31)
      at Promise._settlePromise (node_modules/bluebird/js/release/promise.js:574:18)
      at Promise._settlePromise0 (node_modules/bluebird/js/release/promise.js:619:10)
      at Promise._settlePromises (node_modules/bluebird/js/release/promise.js:699:18)
      at Promise._fulfill (node_modules/bluebird/js/release/promise.js:643:18)
      at MappingPromiseArray.PromiseArray._resolve (node_modules/bluebird/js/release/promise_array.js:126:19)
      at MappingPromiseArray._promiseFulfilled (node_modules/bluebird/js/release/map.js:101:18)
      at Promise._settlePromise (node_modules/bluebird/js/release/promise.js:579:26)
      at Promise._settlePromise0 (node_modules/bluebird/js/release/promise.js:619:10)
      at Promise._settlePromises (node_modules/bluebird/js/release/promise.js:699:18)
      at _drainQueueStep (node_modules/bluebird/js/release/async.js:138:12)
      at _drainQueue (node_modules/bluebird/js/release/async.js:131:9)
      at Async.Object.<anonymous>.Async._drainQueues (node_modules/bluebird/js/release/async.js:147:5)
      at Async.drainQueues (node_modules/bluebird/js/release/async.js:17:14)
 ```

If I update the port to `8888`, against the above commit, I can generate a pactfile.

With the latest commit, using v10.x of pact-js, I need to up both the port to `8888` but also needed to remove the `host` option, or set it to `"127.0.0.1"` as a string.

https://github.com/mmoll/pact_js_update/blob/dc30219da6adda0740462b2769dd0dc35630498e/tests/jest/pact/categories.spec.ts#L9

Generated pact file

```
{
  "consumer": {
    "name": "my_frontend"
  },
  "interactions": [
    {
      "description": "a request to get categories",
      "providerState": "all categories",
      "request": {
        "headers": {
          "Accept": "application/json"
        },
        "method": "GET",
        "path": "/categories"
      },
      "response": {
        "body": [],
        "headers": {
          "Content-Type": "application/json"
        },
        "matchingRules": {
          "$.body": {
            "match": "type"
          }
        },
        "status": 200
      }
    }
  ],
  "metadata": {
    "pact-js": {
      "version": "10.1.1"
    },
    "pactRust": {
      "ffi": "0.3.5",
      "models": "0.4.1"
    },
    "pactSpecification": {
      "version": "2.0.0"
    }
  },
  "provider": {
    "name": "search-service"
  }
}
```

**Comment by YOU54F** [comment]
Also mega kudos for the minimal repro 🙌🏾

**Comment by mmoll** [comment]
Thanks, I think I can work with this. I'm on Linux and with https://github.com/mmoll/pact_js_update/commit/d1d809097fed8e5bb0bcebfc32ad57ce0faca65f it's working locally.

However, AFAICT this means something in the Rust implementation is not IPv6 clean, which should get fixed.

I'll report here in 1-2 days if I can drag the 10.x update through with all the tests in the real code base.

**Comment by mefellows** [comment]
Good pick and thanks for the repro @mmoll!

> With the latest commit, using v10.x of pact-js, I need to change the port to 8888 but also needed to remove the host option, or set it to "127.0.0.1" as a string.

@uglyog is there any implementation advice for Pact JS for IPv6 that I should be aware of? If no user value is provided, we set to `127.0.0.1`, but I think a lot of users will put in `localhost`.

**Comment by rholshausen** [comment]
Localhost should work, unless it is resolving to an IP6 address, in which case I don't think the mock server will respond, because it is using the IP4 loopback.

**Comment by mmoll** [comment]
This issue can be closed, as `127.0.0.1` works.

> Localhost should work, unless it is resolving to an IP6 address, in which case I don't think the mock server will respond, **because it is using the IP4 loopback.**

But why, tho? 🙂

**Comment by rholshausen** [comment]
Historical. The mock server was written in 2016, and that was how things were done then
